### PR TITLE
Add WTF Python Command

### LINKED ISSUE #602: Add Subsection Command for WTFPython
While discussing python-discord/meta#63, a suggestion that came up was to add a command to get a direct link to subsections of the python gotchas list. This can be used to quickly direct people towards relevant sections of this resource.

A possible implementation would fuzzy match the headers on the page, and send the link to the relevant sections.

### PR DESCRIPTION
**Pull request by Shivansh-007**
## Relevant Issues
Closes #602 

## Description
Gets the fuzzy match of a search query, returns `None` if the ratio is below `90` else returns the fuzzy match. If the return is `None`, then we send our custom Error Message else it sends the link to the subsection of the WTF Python README.md.

## Reasoning
Approved by a staff.

## Screenshots
**On Dark Theme**

Screenshot from 2021-03-02 15-22-56

**On Light Theme**

Screenshot from 2021-03-02 15-22-43

(I went blind while teking it, and am helf blind rihgt now, so ignre ayn spellign misteke)

## Did you:

- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] **Lint your code** (`pipenv run lint`)?
- [x] Set the PR to **allow edits from contributors**?

### CODE REVIEW

File: bot/exts/evergreen/wtf_python.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+
+class WTFPython(commands.Cog):
+    """Cog that allows users to retrieve issues from GitHub."""
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+        self.get_wtf_python_readme.start()
+
+        self.data: dict = {}
```
Comment (ChrisLovering) [inline bot/exts/evergreen/wtf_python.py:40]: I don't think you use this anymore.

File: bot/exts/evergreen/wtf_python.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        self.headers: Dict[str] = dict()
+
+    @tasks.loop(hours=1)
+    async def get_wtf_python_readme(self) -> None:
+        """Gets content of README.md of WTF Python Repository."""
```
Comment (ChrisLovering) [inline bot/exts/evergreen/wtf_python.py:45]: Minor grammar.
```suggestion
        """Gets the content of README.md from the WTF Python Repository."""
```

File: bot/exts/evergreen/wtf_python.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+                self.raw = await resp.text()
+                await self.parse_readme(self.raw)
+            else:
+                log.debug(f"Failed to get latest WTF Python README.md. Status code {resp.status}")
+
+    async def parse_readme(self, data: str) -> None:
+        """
+        Parses the README.md into a dict.
+
+        Key is the name of the section and link to it.
```
Comment (ChrisLovering) [inline bot/exts/evergreen/wtf_python.py:57]: Did you rename `Key` to `data`?

File: bot/exts/evergreen/wtf_python.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+            match = re.findall(r"\[▶ (.*)\]\((.*)\)", header)
+            if match:
+                self.headers.update(
+                    {
+                        match[0][0]: f"{BASE_URL}{match[0][1]}"
+                    }
+                )
+
+    def fuzzy_match_header(self, query: str) -> Optional[str]:
+        """Returns the fizzy match of a query if its ratio is above 90 else returns None."""
```
Comment (ChrisLovering) [inline bot/exts/evergreen/wtf_python.py:74]:
```suggestion
        """Returns the fuzzy match of a query if its ratio is above 90 else returns None."""
```

File: bot/exts/evergreen/wtf_python.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        keys = list(self.headers.keys())
+        match = process.extractOne(query, keys)
+        print(match)
```
Comment (ChrisLovering) [inline bot/exts/evergreen/wtf_python.py:77]: Remove debug print, or make it a `log.trace()`

File: bot/exts/evergreen/wtf_python.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            if match:
+                self.headers.update(
+                    {
+                        match[0][0]: f"{BASE_URL}{match[0][1]}"
+                    }
+                )
+
+    def fuzzy_match_header(self, query: str) -> Optional[str]:
+        """Returns the fizzy match of a query if its ratio is above 90 else returns None."""
+        keys = list(self.headers.keys())
```
Comment (ChrisLovering) [inline bot/exts/evergreen/wtf_python.py:75]: Is this wrapper needed? can `process.extractOne()` take `self.headers.keys()` directly?

File: bot/exts/evergreen/wtf_python.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        match = process.extractOne(query, keys)
```
Comment (ChrisLovering) [inline bot/exts/evergreen/wtf_python.py:76]: Maybe unpacking this type into `match, certainty` could help with readability of the if statement below.

**Review by ChrisLovering — CHANGES_REQUESTED** [review]

File: bot/exts/evergreen/wtf_python.py (hunk 3 above)
Comment (Shivansh-007) [reply bot/exts/evergreen/wtf_python.py:57]: No, it is talking about how it parses the `README.md` into `self.headers`

Comment (ChrisLovering) [reply bot/exts/evergreen/wtf_python.py:56]: Ahh, so `self.headers` is a `name: link` dict?

File: bot/exts/evergreen/wtf_python.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+                    {
+                        match[0][0]: f"{BASE_URL}{match[0][1]}"
+                    }
+                )
+
+    def fuzzy_match_header(self, query: str) -> Optional[str]:
+        """Returns the fuzzy match of a query if its ratio is above 90 else returns None."""
+        match, certainty = process.extractOne(
+            query,
+            list(self.headers.keys())
```
Comment (ChrisLovering) [inline bot/exts/evergreen/wtf_python.py:76]: Is the cast to list needed?

File: bot/exts/evergreen/wtf_python.py (hunk 3 above)
Comment (Shivansh-007) [reply bot/exts/evergreen/wtf_python.py:56]: > Ahh, so `self.headers` is a `name: link` dict? If so, could you update this docstring to say that it parses the readme into the `self.headers` dict, where the key is the heading and the value is the link to the heading?

Yes you are right! Alright will update the `doc-string`.

File: bot/exts/evergreen/wtf_python.py (hunk 8 above)
Comment (Shivansh-007) [reply bot/exts/evergreen/wtf_python.py:76]: Yes
```py
d = {'a':1, 'b':2, 'c':3}
print(d.keys())
print(type(d.keys()))
```
Will give:
```py
001 | dict_keys(['a', 'b', 'c'])
002 | <class 'dict_keys'>
```

Comment (ChrisLovering) [reply bot/exts/evergreen/wtf_python.py:76]: Yea, I know that, but as far as I can see `process.extractOne()` will work with a `dict_keys` class directly as its an iterable.

Comment (Shivansh-007) [reply bot/exts/evergreen/wtf_python.py:76]: Ah right,

File: bot/exts/evergreen/wtf_python.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+THUMBNAIL = "https://raw.githubusercontent.com/satwikkansal/wtfpython/master/images/logo.png"
+
+ERROR_MESSAGE = f"""
+Unknown WTF Python Query. Please try to reformulate your query.
+
+**Examples**:
+```md
+{constants.Client.prefix}wtf wild imports
+{constants.Client.prefix}wtf subclass
+{constants.Client.prefix}wtf del
```
Comment (ChrisLovering) [inline bot/exts/evergreen/wtf_python.py:27]: Running both `.wtf del` and `.wtf subclass` results in the error embed, do these need to be updated?

**Review by ChrisLovering — CHANGES_REQUESTED** [review]
Code looks good, just one issue while testing locally. I think it could be a good idea to have a `.wtf list` or similar, to output all of the available pages. Maybe one for after the paginator PR in merged.

File: bot/exts/evergreen/wtf_python.py (hunk 9 above)
Comment (Shivansh-007) [reply bot/exts/evergreen/wtf_python.py:27]: Maybe I could lower down the certainty check.

**Comment by Shivansh-007** [comment]
> Code looks good, just one issue while testing locally. I think it could be a good idea to have a `.wtf list` or similar, to output all of the available pages. Maybe one for after the paginator PR in merged.

Alright will do.

File: bot/exts/evergreen/wtf_python.py [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+                    }
+                )
+
+    def fuzzy_match_header(self, query: str) -> Optional[str]:
+        """Returns the fuzzy match of a query if its ratio is above 90 else returns None."""
+        match, certainty = process.extractOne(query, self.headers.keys())
+        return match if certainty > MINIMUM_CERTAINTY else None
+
+    @staticmethod
+    async def make_embed(query: str, link: str) -> Embed:
```
Comment (ToxicKidz) [inline bot/exts/evergreen/wtf_python.py:82]: Why is this an async function? I don't see it `await`'ing anything.
```suggestion
    def make_embed(query: str, link: str) -> Embed:
```

**Review by ToxicKidz — CHANGES_REQUESTED** [review]

File: bot/exts/evergreen/wtf_python.py (hunk 10 above)
Comment (Xithrius) [reply bot/exts/evergreen/wtf_python.py:82]: I don't think you can create an embed object within a sync function.

Comment (anand2312) [reply bot/exts/evergreen/wtf_python.py:82]: Why not?  Sync function

Comment (Xithrius) [reply bot/exts/evergreen/wtf_python.py:82]: Huh, I guess I was wrong. maybe in previous versions that was a thing? Not sure.

Comment (ToxicKidz) [reply bot/exts/evergreen/wtf_python.py:82]: > I don't think you can create an embed object within a sync function.

Why not?

**Review by ToxicKidz — APPROVED** [review]
LGTM!

File: bot/exts/evergreen/wtf_python.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+    @tasks.loop(hours=1)
+    async def get_wtf_python_readme(self) -> None:
+        """Gets the content of README.md from the WTF Python Repository."""
+        async with self.bot.http_session.get(WTF_PYTHON_RAW_URL) as resp:
+            if resp.status == 200:
+                self.raw = await resp.text()
+                await self.parse_readme(self.raw)
+            else:
+                log.debug(f"Failed to get latest WTF Python README.md. Status code {resp.status}")
```
Comment (RohanJnr) [inline bot/exts/evergreen/wtf_python.py:52]: What happens if the status is not 200?
Then `self.headers` will be an empty dict, then the command fails as line 78 
```python
match, certainty = process.extractOne(query, self.headers.keys())
```
fails with the error
image
and the user does not get any output.

IMO, we should disable to command if we cannot fetch the data from the website/readme or if the response status is not 200 and retry to fetch data for a few times at regular intervals.

File: bot/exts/evergreen/wtf_python.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+        self.get_wtf_python_readme.start()
+
+        self.headers: Dict[str] = dict()
+
+    @tasks.loop(hours=1)
+    async def get_wtf_python_readme(self) -> None:
+        """Gets the content of README.md from the WTF Python Repository."""
+        async with self.bot.http_session.get(WTF_PYTHON_RAW_URL) as resp:
+            if resp.status == 200:
+                self.raw = await resp.text()
```
Comment (RohanJnr) [inline bot/exts/evergreen/wtf_python.py:49]: why is this an instance attribute instead of a local variable?

File: bot/exts/evergreen/wtf_python.py [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+        ext = self.__class__.__name__
+
+        try:
+            action.value(self.bot, ext)
+        except (commands.ExtensionAlreadyLoaded, commands.ExtensionNotLoaded):
+            log.debug(f"Extension `{ext}` is already  is already {verb}ed.")
+        else:
+            log.debug(f"Extension {verb}ed: `{ext}`.")
+
+    async def parse_readme(self, data: str) -> None:
```
Comment (anand2312) [inline bot/exts/evergreen/wtf_python.py:89]: This also doesn't seem to be awaiting anything. Why's it under an `async def`?

**Review by anand2312 — CHANGES_REQUESTED** [review]

File: bot/exts/evergreen/wtf_python.py [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+                        f"Failed to get latest WTF Python README.md on try {x}/{FETCH_TRIES}. Status code {resp.status}"
+                    )
+
+        if failed_tries == 3:
+            action = Action.UNLOAD
+        else:
+            action = Action.LOAD
+
+        verb = action.name.lower()
+        ext = self.__class__.__name__
```
Comment (anand2312) [inline bot/exts/evergreen/wtf_python.py:80]: As for your issue about the extension not being unloaded after too many tries, I think it is possible the error is here - a [Cog](https://discordpy.readthedocs.io/en/latest/ext/commands/cogs.html) _isn't_ an [Extension](https://discordpy.readthedocs.io/en/latest/ext/commands/extensions.html). You're trying to unload the extension by using the cog name, not the extension name (the extension name will be the file name - `exts.evergreen.wtf_python`)

File: bot/exts/evergreen/wtf_python.py [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+        try:
+            action.value(self.bot, ext)
+        except (commands.ExtensionAlreadyLoaded, commands.ExtensionNotLoaded):
+            log.debug(f"Extension `{ext}` is already  is already {verb}ed.")
```
Comment (anand2312) [inline bot/exts/evergreen/wtf_python.py:85]: Grammer here - `f"Extension {ext} is already {verb}ed|`

File: bot/exts/evergreen/wtf_python.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        embed = Embed(
+            title=f"WTF Python Search Result For {query}",
+            colour=constants.Colours.dark_green,
+            description=f"[Go to Repository Section]({link})"
+        )
+        embed.set_thumbnail(url=f"{WTF_PYTHON_RAW_URL}images/logo.png")
+        return embed
+
+    @commands.command(aliases=("wtf",))
+    async def wtf_python(self, ctx: commands.Context, *query: str) -> None:
```
Comment (anand2312) [inline bot/exts/evergreen/wtf_python.py:128]: Was there a reason you chose to use `*query` here instead of using a keyword-only argument (which discord.py uses for a "consume-rest" behaviour)?
```py
async def wtf_python(self, ctx: commands.Context, *, query: str) -> None
```
This will save you the line where you `join` them together. 
https://discordpy.readthedocs.io/en/latest/faq.html#why-do-my-arguments-require-quotes

File: bot/exts/evergreen/wtf_python.py [hunk 17]
```diff
@@ -0,0 +1,151 @@
...
+                    }
+                )
+
+    def fuzzy_match_header(self, query: str) -> Optional[str]:
+        """Returns the fuzzy match of a query if its ratio is above 90 else returns None."""
+        match, certainty = process.extractOne(query, self.headers.keys())
+        return match if certainty > MINIMUM_CERTAINTY else None
+
+    @staticmethod
+    def make_embed(query: str, link: str) -> Embed:
```
Comment (anand2312) [inline bot/exts/evergreen/wtf_python.py:117]: Does this really have to be a function of it's own, when it's used only once?

**Review by anand2312 — APPROVED** [review]
LGTM 👍🏻

**Review by ToxicKidz — CHANGES_REQUESTED** [review]

File: bot/exts/evergreen/wtf_python.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+from discord import Embed
+from discord.ext import commands, tasks
+from fuzzywuzzy import process
+
+from bot import constants
+from bot.bot import Bot
+
+log = logging.getLogger(__name__)
+
+WTF_PYTHON_RAW_URL = "http://raw.githubusercontent.com/satwikkansal/wtfpython/mster/"
```
[the remark this answers is not in the document]
Comment (Shivansh-007) [reply bot/exts/evergreen/wtf_python.py:17]: Whoops, that was a typo.

File: bot/exts/evergreen/wtf_python.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+
+MINIMUM_CERTAINTY = 75
+
+
+class Action(Enum):
+    """Represents an action to perform on an extension."""
+
+    # Need to be partial otherwise they are considered to be function definitions.
+    LOAD = functools.partial(Bot.load_extension)
+    UNLOAD = functools.partial(Bot.unload_extension)
```
[the remark this answers is not in the document]
Comment (Shivansh-007) [reply bot/exts/evergreen/wtf_python.py:41]: `LOAD`, and `UNLOAD` needs to partials as they would be considered as function definitions which is not what we want, if we remove the partials.

File: bot/exts/evergreen/wtf_python.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+
+
+class WTFPython(commands.Cog):
+    """Cog that allows users to retrieve issues from GitHub."""
```
Comment (laundmo) [inline bot/exts/evergreen/wtf_python.py:45]: This docstring seems to be a leftover from copying the cog
```suggestion
    """Cog that allows geting WTF Python entries from the WTF Python repository."""
```

File: bot/exts/evergreen/wtf_python.py [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+            match = re.findall(r"\[▶ (.*)\]\((.*)\)", header)
+            if match:
+                self.headers.update(
+                    {
+                        match[0][0]: f"{BASE_URL}{match[0][1]}"
+                    }
+                )
+
+    def fuzzy_match_header(self, query: str) -> Optional[str]:
+        """Returns the fuzzy match of a query if its ratio is above 90 else returns None."""
```
Comment (laundmo) [inline bot/exts/evergreen/wtf_python.py:113]: Docstring mentions a ratio of 90, though the ratio is a global variable and currently 75
```suggestion
        """Returns the fuzzy match of a query if its ratio is above `MINIMUM_CERTAINTY` else returns None."""
```

**Review by laundmo — CHANGES_REQUESTED** [review]

File: bot/exts/evergreen/wtf_python.py [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+
+                if resp.status == 200:
+                    raw = await resp.text()
+                    self.parse_readme(raw)
+                    log.debug("Successfully fetched the latest WTF Python README.md, breaking out of retry loop")
+                    break
+
+                else:
+                    failed_tries += 1
+                    log.debug(
```
Comment (Kronifer) [inline bot/exts/evergreen/wtf_python.py:70]: Seems to me this should be `log.warning`, but thats just me

File: bot/exts/evergreen/wtf_python.py (hunk 21 above)
Comment (Kronifer) [inline bot/exts/evergreen/wtf_python.py:113]: Could we reword this to "Returns the fuzzy match of a query if its ratio is above 90, otherwise returns None", using else sounds bad

File: bot/exts/evergreen/wtf_python.py [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+
+    def fuzzy_match_header(self, query: str) -> Optional[str]:
+        """Returns the fuzzy match of a query if its ratio is above 90 else returns None."""
+        match, certainty = process.extractOne(query, self.headers.keys())
+        return match if certainty > MINIMUM_CERTAINTY else None
+
+    @commands.command(aliases=("wtf",))
+    async def wtf_python(self, ctx: commands.Context, *, query: str) -> None:
+        """
+        Search wtf python.
```
Comment (Kronifer) [inline bot/exts/evergreen/wtf_python.py:120]: wtf is capitalised in other areas

**Review by Kronifer — CHANGES_REQUESTED** [review]
Few nitpicks, but looks good!

File: bot/exts/evergreen/wtf_python.py (hunk 22 above)
Comment (Shivansh-007) [reply bot/exts/evergreen/wtf_python.py:70]: I am logging a `error` level messages if it fails to fetch it 3 times. I don't see the need to do it for every fetch.

File: bot/exts/evergreen/wtf_python.py (hunk 21 above)
Comment (Shivansh-007) [reply bot/exts/evergreen/wtf_python.py:113]: Explain "bad", I think `else` is good and makes its functionality clear.

**Comment by Shivansh-007** [comment]
I don't think I am having enough time to look into this and it has been stale for quite a while now, so whoever wants can take over the PR, and deliver the feature!

**Review by Xithrius — CHANGES_REQUESTED** [review]
@decorator-factory has noticed that this PR still uses [`fuzzywuzzy`](https://pypi.org/project/fuzzywuzzy/), which [contains the GPLv2 license](https://github.com/seatgeek/fuzzywuzzy/blob/master/LICENSE.txt). We now use [`rapidfuzz`](https://pypi.org/project/rapidfuzz/) because it uses the MIT license, so this new feature needs to reflect this change.

**Comment by brad90four** [comment]
I think I can take this one, looks like rapidfuzz also has an extractOne method and the change from `fuzzy_wuzzy` to `rapidfuzz` is just the import statement.

**Comment by Xithrius** [comment]
@brad90four I shall assign you. `rapidfuzz` is currently what we're going with for the default replacement of `fuzzy_wuzzy`. Be sure to re-lock `poetry.lock`.

**Comment by Xithrius** [comment]
Given that the restructure of this bot occurred, the main file of `bot/exts/evergreen/wtf_python.py` will need to be moved to `bot/exts/utilities/wtf_python.py`.

**Comment by brad90four** [comment]
Should I get perms to Shivansh's repo in order to push changes to the feature branch or should I make a new PR with a branch from my own repo?

**Comment by Xithrius** [comment]
Make a new PR you will.
